### PR TITLE
PLFM-6870: Maintaining Validation Schema Index and Workers For Broadcasting Changes

### PR DESCRIPTION
**Pull request by williamvietnguyen**

### CODE REVIEW

File: services/workers/src/main/resources/validation-schema-index-spb.xml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+						<constructor-arg index="0" ref="countingSemaphore" />
+						<constructor-arg index="1" ref="awsSQSClient" />
+						<constructor-arg index="2">
+							<bean
+								class="org.sagebionetworks.asynchronous.workers.changes.ChangeMessageDrivenWorkerStackConfig">
+								<property name="runner">
+									<bean class="org.sagebionetworks.schema.worker.ValidationSchemaIndexChangeWorker" />
+								</property>
+								<property name="semaphoreLockKey" value="validationSchemaIndexChangeWorker" />
+								<property name="semaphoreMaxLockCount" value="1" />
```
Comment (john-hill) [inline services/workers/src/main/resources/validation-schema-index-spb.xml]: maybe 5 for both counts

File: services/workers/src/main/java/org/sagebionetworks/schema/worker/BroadcastDependantSchemaChangeWorker.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package org.sagebionetworks.schema.worker;
+
+import org.sagebionetworks.asynchronous.workers.changes.ChangeMessageDrivenRunner;
+import org.sagebionetworks.common.util.progress.ProgressCallback;
+import org.sagebionetworks.repo.manager.schema.JsonSchemaManager;
+import org.sagebionetworks.repo.model.message.ChangeMessage;
+import org.sagebionetworks.workers.util.aws.message.RecoverableMessageException;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class BroadcastDependantSchemaChangeWorker implements ChangeMessageDrivenRunner {
```
Comment (john-hill) [inline services/workers/src/main/java/org/sagebionetworks/schema/worker/BroadcastDependantSchemaChangeWorker.java:10]: mock unit test

File: services/workers/src/main/java/org/sagebionetworks/schema/worker/ValidationSchemaIndexChangeWorker.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import org.sagebionetworks.common.util.progress.ProgressCallback;
+import org.sagebionetworks.repo.manager.schema.JsonSchemaManager;
+import org.sagebionetworks.repo.model.message.ChangeMessage;
+import org.sagebionetworks.workers.util.aws.message.RecoverableMessageException;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class ValidationSchemaIndexChangeWorker implements ChangeMessageDrivenRunner {
+
+	@Autowired
+	JsonSchemaManager jsonSchemaManager;
```
Comment (john-hill) [inline services/workers/src/main/java/org/sagebionetworks/schema/worker/ValidationSchemaIndexChangeWorker.java:13]: mock unit test.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java [hunk 4]
```diff
@@ -381,6 +385,7 @@ public void deleteSchemaById(UserInfo user, String $id) {
 			aclDao.canAccess(user, versionInfo.getOrganizationId(), ObjectType.ORGANIZATION, ACCESS_TYPE.DELETE)
 					.checkAuthorizationOrElseThrow();
 		}
+		validationIndexDao.delete(versionInfo.getVersionId());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java]: on cascade delete will handle this.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java [hunk 5]
```diff
@@ -224,15 +228,14 @@ public CreateSchemaResponse createJsonSchema(UserInfo user, CreateSchemaRequest
...
-		if (info.getSemanticVersion() == null) {
-			sendUpdateNotifications(info.getSchemaId());
+		JsonSchema validationSchema = null;
+		if (Boolean.TRUE.equals(request.getDryRun())) {
+			validationSchema = getValidationSchema(info.get$id());
+		} else {
+			// update the validation schema index and send notifications
+			// for bound entities and for dependant schemas
+			validationSchema = createOrUpdateValidationSchemaIndex(info.getVersionId());
 		}
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java:242]: when schema changes fire one dependent event

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java [hunk 6]
```diff
@@ -549,5 +554,42 @@ public JsonSchemaObjectBinding getJsonSchemaObjectBinding(Long objectId, BoundOb
...
+		// semantic version is non-null, then versionId is non-null in dependency table
+		if (info.getSemanticVersion() != null) {
+			versionIds = jsonSchemaDao.getVersionIdsOfDependantsIteratorWithVersionId(versionId);
+		} else {
+			// otherwise grab the versionId's schemaId from the latest version table to access the dependency table
+			String schemaId = info.getSchemaId();
+			versionIds = jsonSchemaDao.getVersionIdsOfDependantsIteratorWithSchemaId(schemaId);
+		}
+		while (versionIds.hasNext()) {
+			transactionalMessenger.sendMessageAfterCommit(versionIds.next(), ObjectType.JSON_SCHEMA, ChangeType.UPDATE);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java]: if the iterator is recursive it will send JSON_SCHEMA events for the full dependency tree.  Let the Database remove duplicates.  The query would use the WITH RECURSIVE feature (https://dev.mysql.com/doc/refman/8.0/en/with.html#common-table-expressions-recursive-examples)
Also write the SQL as a resource and load using :
	public static final String ENTITY_DEPTH_SQL = DDLUtilsImpl
			.loadSQLFromClasspath("sql/EntityDepth.sql");

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImpl.java [hunk 7]
```diff
@@ -557,12 +557,37 @@ public int getBatchSize() {
...
+	public List<String> getNextPageForVersionIdsOfDependantsWithVersionId(String versionId, long limit, long offset) {
+		ValidateArgument.required(versionId, "versionId");
+		return jdbcTemplate.queryForList("SELECT " + COL_JSON_SCHEMA_DEPENDENCY_VERSION_ID + " FROM " 
+				+ TABLE_JSON_SCHEMA_DEPENDENCY + " WHERE "
+				+ COL_JSON_SCHEMA_DEPENDENCY_DEPENDS_ON_VERSION_ID + " = ? LIMIT ? OFFSET ?", 
+				String.class, versionId, limit, offset);
+	}
+	
+	@Override
+	public Iterator<String> getVersionIdsOfDependantsIteratorWithSchemaId(String schemaId) {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImpl.java]: this method can be moved to the manager.

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImpl.java [hunk 8]
```diff
@@ -557,12 +557,37 @@ public int getBatchSize() {
...
-	public List<Long> getVersionIdsOfDependants(String versionId) {
+	public Iterator<String> getVersionIdsOfDependantsIteratorWithVersionId(String versionId) {
 		ValidateArgument.required(versionId, "versionId");
+		return new PaginationIterator<String>((long limit, long offset) -> {
+			return getNextPageForVersionIdsOfDependantsWithVersionId(versionId, limit, offset);
+		}, PAGE_SIZE_LIMIT);
+	}
+	
+	@Override
+	public List<String> getNextPageForVersionIdsOfDependantsWithVersionId(String versionId, long limit, long offset) {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImpl.java]: this goes away

File: lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImpl.java [hunk 9]
```diff
@@ -557,12 +557,37 @@ public int getBatchSize() {
...
+	@Override
+	public Iterator<String> getVersionIdsOfDependantsIteratorWithSchemaId(String schemaId) {
+		ValidateArgument.required(schemaId, "schemaId");
+		return new PaginationIterator<String>((long limit, long offset) -> {
+			return getNextPageForVersionIdsOfDependantsWithSchemaId(schemaId, limit, offset);
+		}, PAGE_SIZE_LIMIT);
+	}
+	
+	@Override
+	public List<String> getNextPageForVersionIdsOfDependantsWithSchemaId(String schemaId, long limit, long offset) {
```
Comment (john-hill) [inline lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImpl.java]: this is recursive.

**Review by john-hill — CHANGES_REQUESTED** [review]

File: lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImplTest.java [hunk 10]
```diff
@@ -1689,4 +1646,70 @@ public void testGetNextPageForEntitiesBoundToSchema() throws JSONObjectAdapterEx
 		assertEquals(results.get(0), secondObjectId);
 		assertEquals(results.get(1), thirdObjectId);
 	}
+
```
Comment (john-hill) [inline lib/jdomodels/src/test/java/org/sagebionetworks/repo/model/dbo/schema/JsonSchemaDaoImplTest.java:1649]: add a test that adds a duplicate (like 3>2>1) and one dependency that is not chained to one.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/EntitySchemaValidator.java [hunk 11]
```diff
@@ -38,8 +39,13 @@ public void validateObject(String entityId) {
 		try {
 			JsonSchemaObjectBinding binding = entityManger.getBoundSchema(entityId);
 			JsonSubject entitySubject = entityManger.getEntityJsonSubject(entityId);
-			JsonSchema validationSchema = jsonSchemaManager
-					.getValidationSchema(binding.getJsonSchemaVersionInfo().get$id());
+			JsonSchema validationSchema = null;
+			try {
+				validationSchema = jsonSchemaManager.getValidationSchemaFromIndex(binding.getJsonSchemaVersionInfo().getVersionId());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/EntitySchemaValidator.java]: the jsonmanager can do this check for everybody.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java [hunk 12]
```diff
@@ -549,5 +554,50 @@ public JsonSchemaObjectBinding getJsonSchemaObjectBinding(Long objectId, BoundOb
...
+	@Override
+	public JsonSchema createOrUpdateValidationSchemaIndex(String versionId, ChangeType changeType) {
+		ValidateArgument.required(versionId, "versionId");
+		JsonSchemaVersionInfo info = jsonSchemaDao.getVersionInfo(versionId);
+		JsonSchema schema = getValidationSchema(info.get$id());
+		validationIndexDao.createOrUpdate(versionId, schema);
+		// update the schemas that depend on this schema when changeType is CREATE (when the schema is created/changed).
+		// ChangeType is UPDATE when we broadcast for dependants, and we do not want to repeat notifications.
+		if (changeType.equals(ChangeType.CREATE)) {
+			transactionalMessenger.sendMessageAfterCommit(versionId, ObjectType.JSON_SCHEMA_DEPENDANT, ChangeType.UPDATE);
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java]: move this to createJsonSchema().  That is the only place where messages of that type.  Then ChangeType can be removed.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java [hunk 13]
```diff
@@ -549,5 +554,50 @@ public JsonSchemaObjectBinding getJsonSchemaObjectBinding(Long objectId, BoundOb
...
+		Iterator<String> versionIds = getVersionIdsOfDependantsIterator(schemaId);
+		while (versionIds.hasNext()) {
+			// send ChangeType.UPDATE so createOrUpdateValidationSchemaIndex doesn't re-send dependant notifications
+			transactionalMessenger.sendMessageAfterCommit(versionIds.next(), 
+					ObjectType.JSON_SCHEMA, ChangeType.UPDATE);
+		}
+	}
+	
+	@Override
+	public JsonSchema getValidationSchemaFromIndex(String versionId) {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java]: getValidationSchemaFromIndex -> getValidationSchema().  This method will also catch not found and rebuild.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java [hunk 14]
```diff
@@ -549,5 +554,50 @@ public JsonSchemaObjectBinding getJsonSchemaObjectBinding(Long objectId, BoundOb
 	public void clearBoundSchema(Long objectId, BoundObjectType objectType) {
 		jsonSchemaDao.clearBoundSchema(objectId, objectType);
 	}
-
+	
+	@Override
+	public JsonSchema createOrUpdateValidationSchemaIndex(String versionId, ChangeType changeType) {
+		ValidateArgument.required(versionId, "versionId");
+		JsonSchemaVersionInfo info = jsonSchemaDao.getVersionInfo(versionId);
+		JsonSchema schema = getValidationSchema(info.get$id());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java]: buildValidationSchema() and is projected.

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java [hunk 15]
```diff
@@ -549,5 +554,50 @@ public JsonSchemaObjectBinding getJsonSchemaObjectBinding(Long objectId, BoundOb
...
+		validationIndexDao.createOrUpdate(versionId, schema);
+		// update the schemas that depend on this schema when changeType is CREATE (when the schema is created/changed).
+		// ChangeType is UPDATE when we broadcast for dependants, and we do not want to repeat notifications.
+		if (changeType.equals(ChangeType.CREATE)) {
+			transactionalMessenger.sendMessageAfterCommit(versionId, ObjectType.JSON_SCHEMA_DEPENDANT, ChangeType.UPDATE);
+		}
+		// If the semantic version is null, notify the entities bound to it, because
+		// schemas with no semantic version are mutable.
+		if (info.getSemanticVersion() == null) {
+			sendUpdateNotifications(info.getSchemaId());
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImpl.java]: sendUpdateNotifications->sendObjectBindingUpdateNotifications

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/schema/EntitySchemaValidatorImplTest.java [hunk 16]
```diff
@@ -83,6 +89,27 @@ public void testValidateObjectWithNotFound() {
 		verify(mockSchemaValidationResultDao).clearResults(entityId, ObjectType.entity);
 	}
 	
+	@Test
+	public void testValidateObjectWithValidationSchemaNotFoundInIndex() {
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/schema/EntitySchemaValidatorImplTest.java]: will move to the manager

**Review by john-hill — CHANGES_REQUESTED** [review]

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImplTest.java [hunk 17]
```diff
@@ -867,7 +875,8 @@ public void testCreateJsonSchemaWithDryRun() {
 				.withJsonSchema(schema).withSemanticVersion(semanticVersionString)
 				.withDependencies(new ArrayList<SchemaDependency>());
 		verify(mockSchemaDao).createNewSchemaVersion(expectedNewSchemaRequest);
-		verify(managerSpy).getValidationSchema(schema.get$id());
+		verify(managerSpy, never()).createOrUpdateValidationSchemaIndex(any());
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImplTest.java:878]: also never send message

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImplTest.java [hunk 18]
```diff
@@ -1624,7 +1640,140 @@ public void testCreateJsonSchemaWithRevalidationMessagesSent() {
...
+	public void testCreateOrUpdateValidationSchemaIndexWithNonNullSemanticVersion() {
+		// we don't want to send notification messages to bound entities here
+		String semanticVersion = "testSemanticVersion";
+		versionInfo.setSemanticVersion(semanticVersion);
+		doReturn(versionInfo).when(mockSchemaDao).getVersionInfo(versionId);
+		doReturn(validationSchema).when(managerSpy).buildValidationSchema(versionInfo.get$id());
+		// call under test
+		JsonSchema result = managerSpy.createOrUpdateValidationSchemaIndex(versionId);
+		verify(mockValidationIndexDao).createOrUpdate(versionId, validationSchema);
+		verify(mockSchemaDao, never()).getObjectIdsBoundToSchemaIterator(any());
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImplTest.java:1700]: never sent message

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImplTest.java [hunk 19]
```diff
@@ -1624,7 +1640,140 @@ public void testCreateJsonSchemaWithRevalidationMessagesSent() {
...
+		// call under test
+		JsonSchema result = managerSpy.getValidationSchema(parsed$Id.toString());
+		assertEquals(result, validationSchema);
+	}
+	
+	@Test
+	public void testGetValidationSchemaWithNotFoundException() {
+		doReturn(versionId).when(managerSpy).getSchemaVersionId(parsed$Id.toString());
+		doThrow(NotFoundException.class).when(mockValidationIndexDao).getValidationSchema(versionId);
+		doReturn(validationSchema).when(managerSpy).createOrUpdateValidationSchemaIndex(versionId);
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/schema/JsonSchemaManagerImplTest.java:1749]: for nonspy use -> when(mock.method()).thenReturn(something)

**Review by john-hill — APPROVED** [review]
